README.rst
Introduction
============

Twisted provides excellent support for creating delayed calls and looping calls,
but when a daemon runs for extended periods of time, it is often necessary to 
schedule tasks to run in a manner that would be more easily described by a
schedule instead of a number of seconds delay. The twisted.scheduling module 
provides an interface for describing a schedule and a single implementation, a
subset of the cron schedules from linux/Unix. In order to use these schedules,
a ScheduledCall class has been created that acts like the 
twisted.internet.task.LoopingCall class. In the future, an extension of the
twisted.application.internet.TimerService may be provided that would use the
ScheduledCall instead of LoopingCall to provide a similar service.

Cron
====

The cron syntax used follows the crontab syntax listed on the `Wikipedia page 
<http://en.wikipedia.org/wiki/Cron>`_ with the exceptions of the names of the 
days of the week which is not supported and the shortcuts which are also not 
supported.


docs/HISTORY.rst
Changelog
=========

1.1 (2011/08/25)
----------------

* Renamed package to TxScheduling

1.0 (2010/06/02)
----------------

* Rename package to twisted.scheduling
* Cleaned up code and added some more error handling
* Cleaned up test code

1.0b1 (2008/12/10)
------------------

* twistedschedule.task.ScheduledCall completed
* twistedschedule.cron.CronSchedule completed
* unittests added for twistedschedule.task.ScheduledCall
* unittests added for twistedschedule.cron


txscheduling/tests/cron.py
from datetime import datetime

from unittest import TestCase, TextTestRunner, TestSuite, TestLoader
from doctest import DocTestSuite

from txscheduling import cron
from txscheduling.cron import CronSchedule, InvalidCronLine



class StarTestCase(TestCase):
    def setUp(self):
        self.schedule = CronSchedule('* * * * *')
    
    def testNextMinute(self):
        """ Next runtime is next minute """
        for i in range(0,59):
            self.assertEqual(self.schedule.getNextEntry(datetime(2008, 01, 01,
                                                                 00, i, 00,
                                                                 00)),
                             datetime(2008,01,01,00,i+1,00,00))
    
    def testNextHour(self):
        """ Next runtime is next hour """
        for i in range(0,23):
            self.assertEqual(self.schedule.getNextEntry(datetime(2008, 01, 01, 
                                                                 i, 59, 00,
                                                                 00)),
                             datetime(2008,01,01,i+1,00,00,00))
    
    def testNextDay(self):
        """ Next runtime is next day """
        for i in range(1,31):
            self.assertEqual(self.schedule.getNextEntry(datetime(2008, 01, i, 
                                                                 23, 59, 00, 
                                                                 00)),
                             datetime(2008,01,i+1,00,00,00,00))
    
    def testNextMonth(self):
        """ Next runtime is next month """
        self.assertEqual(self.schedule.getNextEntry(datetime(2008, 01, 31, 23,
                                                              59, 00, 00)),
                         datetime(2008,02,01,00,00,00,00))
    
    def testNextYear(self):
        """ Next runtime is next year """
        self.assertEqual(self.schedule.getNextEntry(datetime(2008, 12, 31, 23, 
                                                             59, 00, 00)),
                         datetime(2009,01,01,00,00,00,00))

class RangeTestCase(TestCase):
    def setUp(self):
        self.schedule = CronSchedule('15-20 3-6 5-10 5-8 2-3')
    
    def testNextMinute(self):
        """ Test range get next minute """
        # Test minutes that should end up in the current hour at 15 minutes
        for i in range(0,14):
            self.assertEqual(self.schedule.getNextEntry(datetime(2008, 05, 05, 
                                                                 03, i, 00, 
                                                                 00)),
                             datetime(2008,05,05,03,15,00,00))
    
        # Test minutes that should end up in the current hour at i+1 minutes
        for i in range(14,20):
            self.assertEqual(self.schedule.getNextEntry(datetime(2008, 05, 05, 
                                                                 03, i, 00, 
                                                                 00)),
                             datetime(2008,05,05,03,i+1,00,00))
    
        # Test minutes that should end up at hour+1 and 15 minutes
        for i in range(20,60):
            self.assertEqual(self.schedule.getNextEntry(datetime(2008, 05, 05, 
                                                                 03, i, 00, 
                                                                 00)),
                             datetime(2008,05,05,04,15,00,00))
    
    def testNextHour(self):
        """ Test range get next hour """ 
        for i in range(0,3):
            self.assertEqual(self.schedule.getNextEntry(datetime(2008, 05, 05, 
                                                                 i, 59, 00, 
                                                                 00)),
                             datetime(2008,05,05,3,15,00,00))
    
        for i in range(3,6):
            self.assertEqual(self.schedule.getNextEntry(datetime(2008, 05, 05, 
                                                                 i, 59, 00, 
                                                                 00)),
                             datetime(2008,05,05,i+1,15,00,00))
    
        for i in range(6,24):
            self.assertEqual(self.schedule.getNextEntry(datetime(2008, 05, 05, 
                                                                 i, 59, 00, 
                                                                 00)),
                             datetime(2008,05,06,03,15,00,00))
    
    def testNextDay(self):
        """ Test range get next day """
        for i in range(1,5):
            self.assertEqual(self.schedule.getNextEntry(datetime(2008, 05, i, 
                                                                 23, 59, 00, 
                                                                 00)),
                             datetime(2008,05,5,3,15,00,00))
    
        for i in range(5,10):
            self.assertEqual(self.schedule.getNextEntry(datetime(2008, 05, i, 
                                                                 23, 59, 00, 
                                                                 00)),
                             datetime(2008,05,i+1,3,15,00,00))
    
        for i in range(10,13):
            self.assertEqual(self.schedule.getNextEntry(datetime(2008, 05, i, 
                                                                 23, 59, 00, 
                                                                 00)),
                             datetime(2008,05,13,3,15,00,00))
    
        self.assertEqual(self.schedule.getNextEntry(datetime(2008, 05, 13, 23, 
                                                             59, 00, 00)),
                         datetime(2008,05,14,3,15,00,00))
    
        for i in range(14,20):
            self.assertEqual(self.schedule.getNextEntry(datetime(2008, 05, i, 
                                                                 23, 59, 00, 
                                                                 00)),
                             datetime(2008,05,20,3,15,00,00))
    
        self.assertEqual(self.schedule.getNextEntry(datetime(2008, 05, 20, 23, 
                                                             59, 00, 00)),
                         datetime(2008,05,21,3,15,00,00))
    
    def testNextMonth(self):
        """ Test range get next month """
        self.assertEqual(self.schedule.getNextEntry(datetime(2008, 01, 01, 00, 
                                                             00, 00, 00)),
                         datetime(2008,05,05,03,15,00,00))
  
        self.assertEqual(self.schedule.getNextEntry(datetime(2008, 05, 31, 23, 
                                                             59, 00, 00)),
                         datetime(2008,06,03,03,15,00,00))
    
        self.assertEqual(self.schedule.getNextEntry(datetime(2008, 07, 30, 23, 
                                                             59, 00, 00)),
                         datetime(2008,8,05,03,15,00,00))
    
        self.assertEqual(self.schedule.getNextEntry(datetime(2008, 8, 31, 23, 
                                                             59, 00, 00)),
                         datetime(2009,05,05,03,15,00,00))

class AllDOWTestCase(TestCase):
    def setUp(self):
        self.schedule = CronSchedule('*/15 * */5 * *')
  
    def testNextDay(self):
        """ Test all days of the week get next day """
        for i in range(1,30):
            self.assertEqual(self.schedule.getNextEntry(datetime(2008, 01, i, 
                                                                 23, 59, 00, 
                                                                 00)),
                             datetime(2008,01,(i/5)*5+5,00,00,00,00))
  
    def testNextMonth(self):
        """ Test all days of the week get next month """
        self.assertEqual(self.schedule.getNextEntry(datetime(2008, 01, 31, 23, 
                                                             59, 00, 00)),
                         datetime(2008,02,05,00,00,00,00))

class FillingCoverageTestCase(TestCase):
    def test_getFirstDayWithSundayDOW(self):
        """ Test next day with Sunday DOW entry """
        schedule = CronSchedule('* * * * 0,3,5')
    
        self.assertEqual(schedule.getNextEntry(datetime(2008,8,31,23,59,00,00)),
                         datetime(2008,9,3,00,00,00,00))

class AllDOMTestCase(TestCase):
    def setUp(self):
        self.schedule = CronSchedule('*/15 * * * 1,3,5')
  
    def testNextDay(self):
        """ Test all days of the month get next day """
        self.assertEqual(self.schedule.getNextEntry(datetime(2008, 9, 1, 23, 
                                                             59, 00, 00)),
                         datetime(2008,9,3,00,00,00,00))
    
        self.assertEqual(self.schedule.getNextEntry(datetime(2008, 9, 15, 23, 
                                                             59, 00, 00)),
                         datetime(2008,9,17,00,00,00,00))
    
        self.assertEqual(self.schedule.getNextEntry(datetime(2008, 9, 16, 23, 
                                                             59, 00, 00)),
                         datetime(2008,9,17,00,00,00,00))
    
        self.assertEqual(self.schedule.getNextEntry(datetime(2008, 9, 17, 23, 
                                                             59, 00, 00)),
                         datetime(2008,9,19,00,00,00,00))
  
    def testNextMonth(self):
        """ Test all days of the month get next month """
        self.assertEqual(self.schedule.getNextEntry(datetime(2008, 6, 30, 23, 
                                                             59, 00, 00)),
                         datetime(2008,7,2,00,00,00,00))
    
        self.assertEqual(self.schedule.getNextEntry(datetime(2008, 8, 29, 23, 
                                                             59, 00, 00)),
                         datetime(2008,9,1,00,00,00,00))
    
        self.assertEqual(self.schedule.getNextEntry(datetime(2008, 9, 29, 23, 
                                                             59, 00, 00)),
                         datetime(2008,10,1,00,00,00,00))

class SimpleTests(TestCase):
    def setUp(self):
        self.schedule = CronSchedule('* * * * *')
    
    def test_invalid_cronline(self):
        """ CronSchedule invalid cron line """
        self.assertRaises(InvalidCronLine,CronSchedule,' ')
    
    def test_equality(self):
        """ CronSchedule equality testing """
        self.assertEqual(self.schedule,CronSchedule('* * * * *'))
        self.assertEqual(self.schedule == 'blah',False)
    
    def test_invalid_call_to_get_next(self):
        """ CronSchedule getNextEntry with invalid arguments """
        self.assertRaises(ValueError,self.schedule.getNextEntry,'')


def test_suite():
    suite = TestSuite()
    suite.addTest(TestLoader().loadTestsFromTestCase(SimpleTests))
    suite.addTest(TestLoader().loadTestsFromTestCase(AllDOMTestCase))
    suite.addTest(TestLoader().loadTestsFromTestCase(AllDOWTestCase))
    suite.addTest(TestLoader().loadTestsFromTestCase(FillingCoverageTestCase))
    suite.addTest(TestLoader().loadTestsFromTestCase(RangeTestCase))
    suite.addTest(TestLoader().loadTestsFromTestCase(StarTestCase))
    
    suite.addTest(DocTestSuite(cron))
    return suite

if __name__ == '__main__':
    TextTestRunner(verbosity=2).run(test_suite())

txscheduling/tests/__init__.py
import unittest

from txscheduling.tests import cron, task

def test_suite():
    suite = unittest.TestSuite()
    suite.addTests(cron.test_suite())
    suite.addTests(task.test_suite())
    return suite

if __name__ == '__main__':
    unittest.TextTestRunner(verbosity=2).run(test_suite())


txscheduling/task.py
from logging import getLogger

from twisted.python import reflect
from twisted.internet import defer

from txscheduling.interfaces import ISchedule



log = getLogger('twisted.schedule.task')

class ScheduledCall(object):
    """Call a function repeatedly.

    If C{f} returns a deferred, rescheduling will not take place until the
    deferred has fired. The result value is ignored.

    @ivar f: The function to call.
    @ivar a: A tuple of arguments to pass the function.
    @ivar kw: A dictionary of keyword arguments to pass to the function.
    @ivar clock: A provider of
        L{twisted.internet.interfaces.IReactorTime}.  The default is
        L{twisted.internet.reactor}. Feel free to set this to
        something else, but it probably ought to be set *before*
        calling L{start}.

    @type _lastTime: C{float}
    @ivar _lastTime: The time at which this instance most recently scheduled
        itself to run.
    """

    def __init__(self, f, *a, **kw):
        self.call = None
        self.running = False
        self.scheduled = None
        self._lastTime = 0.0
        self.starttime = None
        self.f = f
        self.a = a
        self.kw = kw
        from twisted.internet import reactor
        self.clock = reactor


    def start(self, schedule):
        """Start running function based on the provided schedule.

        @param schedule: An object that provides or can be adapted to an
        ISchedule interface.


        @return: A Deferred whose callback will be invoked with
        C{self} when C{self.stop} is called, or whose errback will be
        invoked when the function raises an exception or returned a
        deferred that has its errback invoked.
        """
        assert not self.running, ("Tried to start an already running "
                                  "ScheduledCall.")
        self.schedule = ISchedule(schedule)
        try:
            self.running = True
            self.deferred = defer.Deferred()
            self.starttime = self.clock.seconds()
            self._lastTime = None
            
            self._reschedule()
            return self.deferred
        except Exception, e:
            log.error('Exception while starting %r: %s' % (self, str(e),))
            self.running = False
            self.deferred = None
            if self.call is not None:
                try:
                    self.call.cancel()
                except:
                    pass
            raise e
        finally:
            log.debug('%r.start(%r) started' % (self, schedule))
        

    def stop(self):
        """ Stop running function. """
        assert self.running, ("Tried to stop a ScheduledCall that was "
                              "not running.")
        self.running = False
        if self.call is not None:
            self.call.cancel()
            self.call = None
            d, self.deferred = self.deferred, None
            d.callback(self)

    def __call__(self):
        def cb(result):
            if self.running:
                self._reschedule()
            else:
                d, self.deferred = self.deferred, None
                d.callback(self)

        def eb(failure):
            self.running = False
            d, self.deferred = self.deferred, None
            d.errback(failure)

        self.call = None
        d = defer.maybeDeferred(self.f, *self.a, **self.kw)
        d.addCallback(cb)
        d.addErrback(eb)


    def _reschedule(self):
        """ Schedule the next iteration of this scheduled call. """
        if self.call is None:
            delay = self.schedule.getDelayForNext()
            self._lastTime = self.clock.seconds() + delay
            self.call = self.clock.callLater(delay, self)


    def __repr__(self):
        if hasattr(self.f, 'func_name'):
            func = self.f.func_name
            if hasattr(self.f, 'im_class'):
                func = self.f.im_class.__name__ + '.' + func
        else:
            func = reflect.safe_repr(self.f)

        return 'ScheduledCall<%s>(%s, *%s, **%s)' % (
            self.schedule, func, reflect.safe_repr(self.a),
            reflect.safe_repr(self.kw))

txscheduling/cron.py
import time
import datetime
import re

import zope.interface

from txscheduling.interfaces import ISchedule



""" This module provides an implementation of the 
txscheduling.interface.ISchedule interface along with two useful 
functions for parsing cron lines. """

class CronSchedule(object):
    zope.interface.implements(ISchedule)
    
    _minutes = None
    _hours = None
    _doms = None # days of the month
    _months = None
    _dows = None # days of the week
  
    def __init__(self,cron_line):
        kwargs = parseCronLine(cron_line)
    
        self._minutes = kwargs.get('minutes')
        self._hours = kwargs.get('hours')
        self._doms = kwargs.get('doms')
        self._months = kwargs.get('months')
        self._dows = kwargs.get('dows')
  
    def __eq__(self,other):
        if not isinstance(other,CronSchedule):
            return False
    
        return (self._minutes == other._minutes and
                self._hours   == other._hours and
                self._doms    == other._doms and
                self._months  == other._months and
                self._dows    == other._dows)
  
    def _getNextMonth(self, current):
        # If the current month is a valid option, try to parse for the next valid day
        if current.month in self._months:
            try:
                return self._getNextDay(current)
            except NoMatch:
                pass
    
        # Find the next month if it occurs in the current year
        for month in self._months:
            if month > current.month:
                return self._getFirstDay(current.replace(month=month,
                                                         day=1,
                                                         hour=self._hours[0],
                                                         minute=self._minutes[0]))
    
        # No remaining months this year, go to next year
        return self._getFirstDay(current.replace(year=current.year+1,
                                                 month=self._months[0],
                                                 day=1,
                                                 hour=self._hours[0],
                                                 minute=self._minutes[0]))
  
    def _getFirstDay(self,current):
        all_doms = False
        all_dows = False
        fdom = current.replace(day=1)
    
        if len(self._doms) == 31:
            all_doms = True
    
        if len(self._dows) == 7:
            all_dows = True
    
        # All days of the week and days of the month, return day = 1
        if all_doms and all_dows:
            return fdom
    
        # All days of the week, return first day of the month
        if all_dows:
            return current.replace(day=self._doms[0])
    
        dows = self._dows
        
        if dows[0] == 0:
            import copy
            dows = copy.copy(dows)
            del dows[0]
            dows.append(7)
    
        current_dow = fdom.isoweekday()
    
        # If the current day of the week is in the days of the week, return
        if not all_dows and current_dow in dows:
            return fdom
    
        # If the first day of the month is specifically listed, return it
        if not all_doms and self._doms[0] == 1:
            return fdom
    
        dow_distance = None
        dom_distance = None
    
        distance = None
    
        if not all_dows:
            for dow in dows:
                if dow > current_dow:
                    dow_distance = dow-current_dow
                    break
      
        if dow_distance is None:
            dow_distance = dows[0]+7-current_dow
    
        if not all_doms:
            dom_distance = self._doms[0] - 1
    
        if not dow_distance is None and not dom_distance is None:
            if dow_distance <= dom_distance:
                distance = dow_distance
            else:
                distance = dom_distance
      
        if distance is None and not dow_distance is None:
            distance = dow_distance
    
        if distance is None and not dom_distance is None:
            distance = dom_distance
    
        return current.replace(day=distance + 1)
  
    def _getNextDay(self, current):
        all_doms = False
        all_dows = False
    
        if len(self._doms) == 31:
            all_doms = True
        
        if len(self._dows) == 7:
            all_dows = True
        
        dows = self._dows
        
        if dows[0] == 0:
            import copy
            dows = copy.copy(dows)
            del dows[0]
            dows.append(7)
        
        # If the current day is a valid option, try to parse for the next valid hour
        if ((all_doms and all_dows) or
            (current.day in self._doms) or
            (current.isoweekday() in dows)):
            try:
                return self._getNextHour(current)
            except NoMatch:
                pass
        
        distance = None
    
        if all_doms and all_dows:
            distance = 1
        else:
            dow_distance = None
            dom_distance = None
          
            if not all_dows:
                current_dow = current.isoweekday()
            
                for dow in dows:
                    if dow > current_dow:
                        dow_distance = dow-current_dow
                        break
                
                if dow_distance is None:
                    dow_distance = dows[0]+7-current_dow
          
            if not all_doms:
                for dom in self._doms:
                    if dom > current.day:
                        dom_distance = dom - current.day
                        break
          
            if not dow_distance is None and dom_distance is not None:
                if dow_distance <= dom_distance:
                    distance = dow_distance
                else:
                    distance = dom_distance
          
            if distance is None and dow_distance is not None:
                distance = dow_distance
          
            if distance is None and dom_distance is not None:
                distance = dom_distance
          
            if distance is None or distance <= 0:
                raise NoMatch('no matching days in month')
          
        next_day = (current.replace(hour=self._hours[0],
                                    minute=self._minutes[0]) + 
                    distance * datetime.timedelta(days=1))
        
        if next_day.month != current.month:
          raise NoMatch('no remaining days in the current month')
        
        return next_day
  
    def _getNextHour(self, current):
        if current.hour in self._hours:
            try:
                return self._getNextMinute(current)
            except NoMatch:
                pass
    
        for hour in self._hours:
            if hour > current.hour:
                return current.replace(hour=hour,minute=self._minutes[0])
        
        raise NoMatch('no remaining hours in the current day')
  
    def _getNextMinute(self, current):
        for minute in self._minutes:
            if minute > current.minute:
                return current.replace(minute=minute)
    
        raise NoMatch('no remaining minutes in the current hour')
  
    def getNextEntry(self,current=None):
        if current is None:
            current = datetime.datetime.now()
        
        if not isinstance(current,datetime.datetime):
            raise ValueError('current value must be a datetime.datetime object')
        
        return self._getNextMonth(current.replace(second=0,microsecond=0))
    
    def getDelayForNext(self):
        next = self.getNextEntry()
        
        return time.mktime(next.timetuple()) - time.time()


class InvalidCronLine(Exception):
  pass

class InvalidCronEntry(Exception):
  pass

class NoMatch(Exception):
  pass

_cronStepRe = re.compile('^\*/(?P<step>\d{1,2})$')
_cronRangeRe = re.compile('^(?P<begin>\d{1,2})-(?P<end>\d{1,2})$')
_cronRangeStepRe = re.compile('^(?P<begin>\d{1,2})-(?P<end>\d{1,2})/(?P<step>\d{1,2})$')

def parseCronLine(line):
    """
    Parse a standard cron string (minus the command) and return them as a
    dictionary. The syntax for this was pulled from the
    Wikipedia page: http://en.wikipedia.org/wiki/Cron
    Currently, there is no support for textual days of the week
    (i.e. Monday,Tuesday).
    
    Examples:
    
    >>> parseCronLine('* * * * *') == {
    ...     'hours': range(0,24),
    ...     'doms': range(1,32),
    ...     'minutes': range(0,60),
    ...     'dows': range(0,7),
    ...      'months': range(1,13)}
    True
    
    Cron lines must have 5 whitespace separated entries
    >>> parseCronLine('* * * *')
    Traceback (most recent call last):
    ...
    InvalidCronLine: Improper number of elements encountered: 4
    
    Cron lines cannot be empty
    >>> parseCronLine('')
    Traceback (most recent call last):
    ...
    InvalidCronLine: Empty cron line provided
    
    Cron lines must be a string
    >>> parseCronLine(True)
    Traceback (most recent call last):
    ...
    InvalidCronLine: Cron line must be a string
    
    """
    if not line:
        raise InvalidCronLine('Empty cron line provided')
    
    if not isinstance(line, basestring):
        raise InvalidCronLine('Cron line must be a string')
    
    line = re.split('\s+',line.strip())
    
    if len(line) != 5:
        raise InvalidCronLine('Improper number of elements encountered: %s' % len(line)) 
      
    schedule = {}
      
    schedule['minutes'] = parseCronEntry(line[0],0,59)
    schedule['hours']   = parseCronEntry(line[1],0,23)
    schedule['doms']    = parseCronEntry(line[2],1,31)
    schedule['months']  = parseCronEntry(line[3],1,12)
    schedule['dows']    = parseCronEntry(line[4],0,6)
      
    return schedule

def parseCronEntry(entry,min,max):
    """Parse a single cron entry for something like hours or minutes from a cron
    scheduling line.  The given min and max are used to verify that results are
    in the proper range. The following formats are supports:
    * => All in the available range
    */5 => Only those values in the available range that are divisible by five
    1-5 => The range of 1-5
    And any combination of the above using commas to separate the entries.
    
    Examples:
    
    A single entry
    >>> parseCronEntry('0',0,12)
    [0]
    
    A star entry
    >>> parseCronEntry('*',0,3)
    [0, 1, 2, 3]
    
    A range entry
    >>> parseCronEntry('3-5',0,12)
    [3, 4, 5]
    
    A frequency entry
    >>> parseCronEntry('*/3',0,12)
    [0, 3, 6, 9, 12]
    
    A frequency that doesn't match the minimum
    >>> parseCronEntry('*/3',1,12)
    [3, 6, 9, 12]
    
    A ranged frequency entry
    >>> parseCronEntry('3-9/3',0,12)
    [3, 6, 9]
    
    A ranged frequency entry where frequency doesn't match the range
    >>> parseCronEntry('2-10/3',0,12)
    [3, 6, 9]
    
    A set of entries
    >>> parseCronEntry('1,3,5',1,12)
    [1, 3, 5]
    
    A set of ranges
    >>> parseCronEntry('1-3,6-9',1,12)
    [1, 2, 3, 6, 7, 8, 9]
    
    A complex entry
    >>> parseCronEntry('*/5,1,12-14,22-28/3',1,30)
    [1, 5, 10, 12, 13, 14, 15, 20, 24, 25, 27, 30]
    
    Minimum argument must be convertible to an integer
    >>> parseCronEntry('1','a', 5000)
    Traceback (most recent call last):
      ...
    ValueError: minimum and maximum should be convertible to integers
    
    Maximum argument must be convertible to an integer
    >>> parseCronEntry('1',5, 'a')
    Traceback (most recent call last):
      ...
    ValueError: minimum and maximum should be convertible to integers
    
    Minimum argument must not be negative
    >>> parseCronEntry('1',-100,100)
    Traceback (most recent call last):
      ...
    ValueError: minimum must be non-negative
    
    Minimum argument must be less than the maximum argument
    >>> parseCronEntry('1',1,0)
    Traceback (most recent call last):
      ...
    ValueError: minimum must be less than or equal to maximum
    
    Entries are required
    >>> parseCronEntry(None, 1, 5)
    Traceback (most recent call last):
      ...
    InvalidCronEntry: Empty cron entry
    
    >>> parseCronEntry('', 1, 5)
    Traceback (most recent call last):
      ...
    InvalidCronEntry: Empty cron entry
    
    Entries must be a string
    >>> parseCronEntry([1], 1, 5)
    Traceback (most recent call last):
      ...
    InvalidCronEntry: Cron entry must be a string
    
    Entries must be inside of the allowed range
    >>> parseCronEntry('0', 1, 5)
    Traceback (most recent call last):
      ...
    InvalidCronEntry: Value, 0-0, out of allowed range: 1-5
        
    >>> parseCronEntry('6', 1, 5)
    Traceback (most recent call last):
      ...
    InvalidCronEntry: Value, 6-6, out of allowed range: 1-5
    
    >>> parseCronEntry('-6', 1, 5)
    Traceback (most recent call last):
      ...
    InvalidCronEntry: Value, -6--6, out of allowed range: 1-5
    
    Ranges must be in minimum-maximum order
    >>> parseCronEntry('3-1', 1, 5)
    Traceback (most recent call last):
      ...
    InvalidCronEntry: Invalid range or step specified: 3-1
    
    Range frequency must be valid for range size
    >>> parseCronEntry('1-5/40', 1, 5)
    Traceback (most recent call last):
      ...
    InvalidCronEntry: Invalid range or step specified: 1-5/40
    
    Negative frequencies are not allowed
    >>> parseCronEntry('1-5/-2', 1, 5)
    Traceback (most recent call last):
      ...
    InvalidCronEntry: Invalid cron entry
    
    >>> parseCronEntry('this is garbage',1,5)
    Traceback (most recent call last):
      ...
    InvalidCronEntry: Invalid cron entry
    
    """
      
    if not entry:
        raise InvalidCronEntry('Empty cron entry')
    
    if not isinstance(entry, basestring):
        raise InvalidCronEntry('Cron entry must be a string')
    
    try:
        min = int(min)
        max = int(max)
    except ValueError:
        raise ValueError('minimum and maximum should be convertible to integers')
      
    if min > max:
        raise ValueError('minimum must be less than or equal to maximum')
      
    if min < 0:
        raise ValueError('minimum must be non-negative')
      
    try:
        entry = entry.split(',')
    except AttributeError:
        raise InvalidCronEntry('Cron entry must be a string')
      
    total = set()
      
    for e in entry:
        try:
            int_val = int(e)
        except ValueError:
            pass
        else:
            total.add(int_val)
            continue
        
        begin = None
        end = None
        step = 1
        
        if e == '*':
            begin = min
            end = max + 1
        
        if begin is None:
            #If this match works, then it is of the form */int
            match = _cronStepRe.search(e)
        
            if not match is None:
                begin = min
                end = max + 1
                step = int(match.group('step'))
            
        if begin is None:
            match = _cronRangeRe.search(e)
        
            if not match is None:
                begin = int(match.group('begin'))
                end = int(match.group('end')) + 1
                step = 1
        
        if begin is None:
            match = _cronRangeStepRe.search(e)
          
            if not match is None:
                begin = int(match.group('begin'))
                end = int(match.group('end')) + 1
                step = int(match.group('step'))
        
        if (begin is not None and begin < end and step > 0 and 
            begin >= min and end <= max + 1):
            
            # need to align the start properly
            while begin % step != 0 and begin < end:
                begin += 1
          
            if begin == end and begin % step != 0:
                raise InvalidCronEntry('Invalid range or step specified: %s' % (e))
          
            total.update(range(begin,end,step))
        elif not begin is None:
            raise InvalidCronEntry('Invalid range or step specified: %s' % (e))
      
    if len(total) == 0:
        raise InvalidCronEntry('Invalid cron entry')
      
    total = list(total)
    total.sort()
      
    if total[0] < min or total[len(total)-1] > max:
        raise InvalidCronEntry('Value, %s-%s, out of allowed range: %s-%s' % (total[0],total[len(total)-1],min,max))
      
    return total

__all__ = [
    'CronSchedule',
    'InvalidCronLine'
]